### PR TITLE
Add 'bot', 'minor', 'prepend', 'nocreate' and 'createonly' options, and fix issue with watchlist

### PR DESCRIPTION
**Pull request by marijnvanwezel**
This pull request adds a number of API options, namely:

- `bot`: to mark the edit as a bot edit;
- `minor`: to mark the edit as minor;
- `prepend`: to prepend the text to the current page;
- `nocreate`: to disable page creation if the page does not exist yet;
- `createonly`: to disable editing if the page already exists.

It also fixes an issue with the `watchlist` option (which simply didn't do what it was supposed to), and makes its behaviour identical to that of the `edit` API.

### CODE REVIEW

File: src/WSSlots.php [hunk 1]
```diff
@@ -64,9 +96,14 @@ final public static function editSlots(
 		User $user,
 		WikiPage $wikiPage,
 		array $slotUpdates,
-		string $summary,
+		string $summary = '',
 		bool $append = false,
-		string $watchlist = ""
+		bool $prepend = false,
+		string $watchlist = "",
```
Comment (wgevaert) [inline src/WSSlots.php:102]: Note: This is a backwards-compatibility break: prepend arg is where watchlist used to be.

File: src/WSSlots.php [hunk 2]
```diff
@@ -166,14 +215,36 @@ final public static function editSlots(
 		$flags = EDIT_INTERNAL;
 		$comment = CommentStoreComment::newUnsavedComment( $summary );
 
-		if ( $watchlist === "nochange" ) {
```
Comment (wgevaert) [inline src/WSSlots.php:169]: Preferably there still is the option to add EDIT_SUPPRESS_RC for other types of edits.

File: src/WSSlots.php [hunk 3]
```diff
@@ -204,4 +275,48 @@ final public static function getSlotContent( WikiPage $wikiPage, string $slot ):
...
+	 * @return bool
+	 */
+	protected static function getWatchlistValue(
+		string $watchlist,
+		Title $title,
+		User $user,
+	): bool {
+		$services = MediaWikiServices::getInstance();
+
+		$userWatching = $services->getWatchlistManager()->isWatchedIgnoringRights( $user, $title );
```
Comment (wgevaert) [inline src/WSSlots.php:294]: It does not make sense to directly use `$services->getWatchlistManager()` here, but a bit higher up only call it with `method_exists( MediaWikiServices::class, 'getWatchlistManager' )` around it.

**Review by wgevaert — CHANGES_REQUESTED** [review]
Looks good overall, some nitpicks

File: src/API/ApiEditSlot.php [hunk 4]
```diff
@@ -44,11 +44,17 @@ public function execute() {
 			$params["slot"],
 			$params["summary"],
 			$params["append"],
-			$params["watchlist"]
+			$params["prepend"],
```
Comment (wgevaert) [inline src/API/ApiEditSlot.php:47]: I think we want watchlist before prepend if you have it in that order below.

Now that I think of it, when we do make it "Backwards compatible", shouldn't we set "suppress" as part of "watchlist"?

I mean, previously this:
watchlist=nochange
Would give what now is this:
suppress=true

Not the same at all.
So maybe it does make sense to break backwards compatibility on purpose? I am not sure.

File: src/WSSlots.php [hunk 5]
```diff
@@ -166,14 +220,40 @@ final public static function editSlots(
...
+			// <=1.36
+			$watchedItemStore = MediaWikiServices::getInstance()->getWatchedItemStore();
+
+			if ( $watch ) {
+				$watchedItemStore->addWatch( $user, $titleObject );
+			} else {
+				$watchedItemStore->removeWatch( $user, $titleObject );
+			}
+
+			$user->invalidateCache();
```
Comment (wgevaert) [inline src/WSSlots.php:254]: I think "invalidateCache" should not be called when `$watchlist=nochange`?
There is no change, no need to invalidate cache.

Or did I misunderstand what invalidatecache does? (I have no idea tbh)

Comment (marijnvanwezel) [reply src/WSSlots.php:254]: I copied this from core :)

File: src/API/ApiEditSlot.php (hunk 4 above)
Comment (marijnvanwezel) [reply src/API/ApiEditSlot.php]: I don't consider fixing an issue that was clearly wrong (setting `watchlist` did not do *anything* with watchlists) a BC break.

**Pull request merged**